File: platform/src/java/com/xcase/rest/generator/swagger/SwaggerJavaProxyGenerator.java
package com.xcase.rest.generator.swagger;

import com.xcase.rest.generator.RESTServiceDefinition;
import com.xcase.rest.generator.RESTApiProxySettingsEndpoint;
import com.xcase.rest.generator.RESTApiProxySettings;
import com.xcase.rest.generator.IProxyGenerator;
import com.xcase.rest.generator.IAPIProxySettingsEndpoint;
import com.xcase.rest.generator.ClassDefinition;
import com.xcase.rest.generator.ProxyDefinition;
import com.xcase.rest.generator.TypeDefinition;
import com.xcase.rest.generator.Parameter;
import com.xcase.rest.generator.RESTParser;
import com.xcase.rest.generator.ParameterIn;
import com.xcase.rest.generator.Operation;
import com.google.gson.*;
import java.lang.invoke.MethodHandles;
import java.io.*;
import java.nio.file.*;
import java.util.*;
import org.apache.logging.log4j.LogManager;
import org.apache.logging.log4j.Logger;

public class SwaggerJavaProxyGenerator extends SwaggerProxyGenerator implements IProxyGenerator
{
    /**
     * log4j object.
     */
    protected static final Logger LOGGER = LogManager.getLogger(MethodHandles.lookup().lookupClass());
    
    public SwaggerJavaProxyGenerator(String packageName) {
        this.packageName = packageName;
        this.objectsPackageName = this.packageName + ".objects";
        this.objectsPath = this.objectsPackageName.replace(".", "/");
    }

    @Override
    public RESTServiceDefinition generateSourceString(IAPIProxySettingsEndpoint swaggerApiProxySettingsEndPoint, String swaggerDocument) throws Exception {
        LOGGER.debug("starting generateSourceString()");
        return null; 
    }

    @Override
    public RESTServiceDefinition generateSourceString(String swaggerDocument) throws Exception {
        LOGGER.debug("starting GenerateSourceString()");
        try {
            endpointHashMap = new HashMap<IAPIProxySettingsEndpoint, String>();
            RESTApiProxySettingsEndpoint swaggerApiProxySettingsEndPoint = new RESTApiProxySettingsEndpoint();
            swaggerApiProxySettingsEndPoint.AppendAsyncToMethodName = false;
            swaggerApiProxySettingsEndPoint.Namespace = "com.xcase.integrate.objects";
            endpointHashMap.put(swaggerApiProxySettingsEndPoint, swaggerDocument);
            LOGGER.debug("about to process REST document");
            return processSwaggerDocuments();
        }
        catch (Exception e) {
            LOGGER.warn("exception generating source string: " + e.getMessage());
            throw e;
        }
    }
    
    public List<RESTServiceDefinition> generateSourceStringForRestServiceDefinitionList(IAPIProxySettingsEndpoint swaggerApiProxySettingsEndPoint, String swaggerDocument, String username, String password, String tenant) throws Exception {
        LOGGER.debug("starting generateSourceString()");
        try {
            endpointHashMap = new HashMap<IAPIProxySettingsEndpoint, String>();
            ((RESTApiProxySettingsEndpoint)swaggerApiProxySettingsEndPoint).AppendAsyncToMethodName = false;
            endpointHashMap.put(swaggerApiProxySettingsEndPoint, swaggerDocument);
            LOGGER.debug("about to process REST document");
            List<RESTServiceDefinition> restServiceDefinitionList = processSwaggerDocumentsForRestServicDefinitionList(objectsPackageName, username, password, tenant);
            LOGGER.debug("processed REST document");
            return restServiceDefinitionList;
        }
        catch (Exception e) {
            LOGGER.warn("exception generating source string: " + e.getMessage());
            throw e;
        }
    }

    @Override
    public RESTServiceDefinition generateSourceString(IAPIProxySettingsEndpoint swaggerApiProxySettingsEndPoint, String swaggerDocument, String username, String password, String tenant) throws Exception {
        LOGGER.debug("starting generateSourceString()");
        try {
            endpointHashMap = new HashMap<IAPIProxySettingsEndpoint, String>();
            ((RESTApiProxySettingsEndpoint)swaggerApiProxySettingsEndPoint).AppendAsyncToMethodName = false;
            endpointHashMap.put(swaggerApiProxySettingsEndPoint, swaggerDocument);
            LOGGER.debug("about to process REST document");
            RESTServiceDefinition restServiceDefinition = processSwaggerDocuments(username, password, tenant);
            LOGGER.debug("processed REST document");
            return restServiceDefinition;
        }
        catch (Exception e) {
            LOGGER.warn("exception generating source string: " + e.getMessage());
            throw e;
        }
    }

    @Override
    public RESTServiceDefinition generateSourceString(IAPIProxySettingsEndpoint[] endpoints) throws Exception {
        LOGGER.debug("starting generateSourceString()");
        try {
            endpointHashMap = new HashMap<IAPIProxySettingsEndpoint, String>();
            LOGGER.debug("requesting REST documents...");
            for (IAPIProxySettingsEndpoint endPoint : endpoints) {
                String requestUri = endPoint.getUrl();
                LOGGER.debug("requested: " + requestUri);
                getEndpointDoc(requestUri, endPoint);
            }

            LOGGER.debug("waiting for REST documents to complete downloading...");
            return processSwaggerDocuments();
        }
        catch (Exception e) {
            LOGGER.warn("exception generating source string: " + e.getMessage());
            throw e;
        }
    }
    
    private static List<RESTServiceDefinition> processSwaggerDocumentsForRestServicDefinitionList(String objectsPackageName, String username, String password, String tenant) throws Exception {
        LOGGER.debug("starting processSwaggerDocuments()");
        List<RESTServiceDefinition> swaggerServiceDefinitionList = new ArrayList<RESTServiceDefinition>();
        for (Map.Entry<IAPIProxySettingsEndpoint, String> swaggerDocDictionaryEntry : endpointHashMap.entrySet()) {
            IAPIProxySettingsEndpoint endPoint = swaggerDocDictionaryEntry.getKey();
            LOGGER.debug("processing " + endPoint.getUrl());
            String methodNameAppend = "";
            if (endPoint.getAppendAsyncToMethodName()) {
                methodNameAppend = "Async";
            }

            String result = swaggerDocDictionaryEntry.getValue();
            SwaggerParser parser = new SwaggerParser();
            RESTServiceDefinition restServiceDefinition = parser.parseDocForRestServiceDefinition(result, (RESTApiProxySettingsEndpoint)endPoint, objectsPackageName);
            restServiceDefinition.writeToStringBuilders();
            swaggerServiceDefinitionList.add(restServiceDefinition);
        }
        
        LOGGER.debug("finishing processSwaggerDocuments()");
        return swaggerServiceDefinitionList;
    }

    private static RESTServiceDefinition processSwaggerDocuments(String username, String password, String tenant) throws Exception {
//        LOGGER.debug("starting processSwaggerDocuments()");
        RESTServiceDefinition swaggerServiceDefinition = new RESTServiceDefinition();
        List<String> sourceStringList = new ArrayList<String>();
        for (Map.Entry<IAPIProxySettingsEndpoint, String> swaggerDocDictionaryEntry : endpointHashMap.entrySet()) {
            IAPIProxySettingsEndpoint endPoint = swaggerDocDictionaryEntry.getKey();
            LOGGER.debug("processing " + endPoint.getUrl());
            String methodNameAppend = "";
            if (endPoint.getAppendAsyncToMethodName()) {
                methodNameAppend = "Async";
            }

            String result = swaggerDocDictionaryEntry.getValue();
            SwaggerParser parser = new SwaggerParser();
            ProxyDefinition proxyDefinition = parser.parseDoc(result, (RESTApiProxySettingsEndpoint)endPoint);
            String endPointString = "http://" + proxyDefinition.Host + proxyDefinition.BasePath;
            if (!endPointString.endsWith("/")) {
                endPointString = endPointString + "/";
            }

            swaggerServiceDefinition.EndPoint = endPointString;
            List<Operation> operationList = proxyDefinition.Operations;
            List<String> proxies = new ArrayList<String>();
            for (Operation operation : operationList) {
                if (!proxies.contains(operation.ProxyName)) {
                    proxies.add(operation.ProxyName);
                }
            }
            
            /* Write interface for each proxy in proxies */
            for (String proxy : proxies) {
//                LOGGER.debug("next proxy " + proxy);
                StringBuilder interfaceStringBuilder = writeInterfaceStringBuilderForProxy(proxyDefinition, proxy, endPoint, methodNameAppend);
//                LOGGER.debug("created interfaceStringBuilder for " + proxy);
                sourceStringList.add(interfaceStringBuilder.toString());
            }

            /* Write class for each proxy in proxies */
            for (String proxy : proxies) {
//                LOGGER.debug("next proxy " + proxy);
                String className = SwaggerParser.fixTypeName(proxy) + "WebProxy";
//                LOGGER.debug("proxy className is " + className);
                swaggerServiceDefinition.ProxyClasses.put(className, proxyDefinition);
                StringBuilder classStringBuilder = writeProxyStringBuilderForProxy(proxyDefinition, proxy, endPoint, methodNameAppend);
//                LOGGER.debug("created classStringBuilder for " + proxy);
                sourceStringList.add(classStringBuilder.toString());
            }

            /* Model Classes */
            for (ClassDefinition classDefinition : proxyDefinition.ClassDefinitions) {
                StringBuilder modelClassStringBuilder = writeClassDefinitionToStringBuilder(classDefinition, endPoint);
                sourceStringList.add(modelClassStringBuilder.toString());
            }
        }

        swaggerServiceDefinition.SourceStrings = sourceStringList.toArray(new String[0]);
        return swaggerServiceDefinition;
    }

    private static RESTServiceDefinition processSwaggerDocuments() throws Exception {
//        LOGGER.debug("starting processSwaggerDocuments()");
        return processSwaggerDocuments("Admin", "1nt@ppC10ud2016", "tenant1");
    }

    public static RESTApiProxySettings getSettings(String path) throws Exception {
        String value = new String(Files.readAllBytes(Paths.get(path)));
        return new Gson().fromJson(value, RESTApiProxySettings.class);
    }
}

File: dotnetcore/XCaseServiceClient/XCase.Swagger.ProxyGenerator/Generator/SwaggerJavaProxyGenerator.cs
namespace XCase.REST.ProxyGenerator.Generator
{
    using System;
    using System.Collections.Concurrent;
    using System.Collections.Generic;
    using System.IO;
    using System.Linq;
    using System.Net;
    using System.Net.Http;
    using System.Reflection;
    using System.Security;
    using System.Text;
    using System.Threading.Tasks;
    using Microsoft.CSharp;
//    using Microsoft.Owin.Testing;
    using Microsoft.Extensions.Logging;
    using Newtonsoft.Json;
    using Serilog;
    using Serilog.Events;
    using XCase.ProxyGenerator;
    using XCase.ProxyGenerator.REST;
    using XCase.REST.ProxyGenerator.OpenAPI;

    public class SwaggerJavaProxyGenerator : JavaProxyGenerator
    {
        #region Logger Setup

        /// <summary>
        /// A log4net log instance.
        /// </summary>
        //static readonly Serilog.ILogger Log = new LoggerConfiguration().MinimumLevel.Debug().WriteTo.File("XCaseServiceClient.log").WriteTo.Console(restrictedToMinimumLevel: LogEventLevel.Information).CreateLogger();

        #endregion

        public override IServiceDefinition GenerateSourceString(IAPIProxySettingsEndpoint swaggerApiProxySettingsEndPoint, string swaggerDocument)
        {
            Log.Debug("starting GenerateSourceString()");
            return null; 
        }

        public override IServiceDefinition GenerateSourceString(string swaggerDocument)
        {
            Log.Debug("starting GenerateSourceString()");
            try
            {
                swaggerDocDictionary = new ConcurrentDictionary<IAPIProxySettingsEndpoint, string>();
                SourceStringBuilder = new StringBuilder();
                RESTApiProxySettingsEndPoint swaggerApiProxySettingsEndPoint = new RESTApiProxySettingsEndPoint();
                swaggerApiProxySettingsEndPoint.AppendAsyncToMethodName = false;
                swaggerApiProxySettingsEndPoint.Namespace = "com.xcase.integrate.objects";
                swaggerDocDictionary.GetOrAdd(swaggerApiProxySettingsEndPoint, swaggerDocument);
                Log.Debug("about to process REST document");
                return ProcessSwaggerDocuments();
            }
            catch (AggregateException ae)
            {
                Log.Warning("aggregate exception generating source string: " + ae.Message);
                throw;
            }
            catch (Exception e)
            {
                Log.Warning("exception generating source string: " + e.Message);
                throw;
            }
        }

        public override IServiceDefinition GenerateSourceString(IAPIProxySettingsEndpoint swaggerApiProxySettingsEndPoint, string swaggerDocument, string username, string password, string tenant)
        {
            Log.Debug("starting GenerateSourceString()");
            try
            {
                swaggerDocDictionary = new ConcurrentDictionary<IAPIProxySettingsEndpoint, string>();
                SourceStringBuilder = new StringBuilder();
                //SwaggerApiProxySettingsEndPoint swaggerApiProxySettingsEndPoint = new SwaggerApiProxySettingsEndPoint();
                ((RESTApiProxySettingsEndPoint)swaggerApiProxySettingsEndPoint).AppendAsyncToMethodName = false;
                swaggerDocDictionary.GetOrAdd(swaggerApiProxySettingsEndPoint, swaggerDocument);
                Log.Debug("about to process REST document");
                RESTServiceDefinition restServiceDefinition = ProcessSwaggerDocuments(username, password, tenant);
                Log.Debug("processed REST document");
                return restServiceDefinition;
            }
            catch (AggregateException ae)
            {
                Log.Warning("aggregate exception generating source string: " + ae.Message);
                throw;
            }
            catch (Exception e)
            {
                Log.Warning("exception generating source string: " + e.Message);
                throw;
            }
        }

        public override IServiceDefinition GenerateSourceString(IAPIProxySettingsEndpoint[] endpoints)
        {
            Log.Debug("starting GenerateSourceString()");
            try
            {
                swaggerDocDictionary = new ConcurrentDictionary<IAPIProxySettingsEndpoint, string>();
                SourceStringBuilder = new StringBuilder();
                List<Task<string>> taskList = new List<Task<string>>();
                foreach (IAPIProxySettingsEndpoint endPoint in endpoints)
                {
                    string requestUri = endPoint.GetUrl();
                    Log.Debug("about to add task for {0}", requestUri);
                    Task<string> endpointTask = GetEndpointDoc(requestUri);
                    taskList.Add(endpointTask);
                    Log.Debug("added endpointTask");
                    string swaggerString = endpointTask.Result;
                    Log.Debug("swaggerString is {0}", swaggerString);
                    swaggerDocDictionary.GetOrAdd(endPoint, swaggerString);
                }

                Log.Debug("waiting for REST documents to complete downloading...");
                Task.WhenAll(taskList);
                Log.Debug("REST documents completed downloading");
                return ProcessSwaggerDocuments();
            }
            catch (AggregateException ae)
            {
                Log.Warning("aggregate exception generating source string: " + ae.Message);
                throw;
            }
            catch (Exception e)
            {
                Log.Warning("exception generating source string: " + e.Message);
                throw;
            }
        }

        private static RESTServiceDefinition ProcessSwaggerDocuments(string username, string password, string tenant)
        {
            Log.Debug("starting ProcessSwaggerDocuments()");
            RESTServiceDefinition swaggerServiceDefinition = new RESTServiceDefinition();
            List<string> sourceStringList = new List<string>();
            foreach (KeyValuePair<IAPIProxySettingsEndpoint, string> swaggerDocDictionaryEntry in swaggerDocDictionary.OrderBy(x => x.Key.GetId()))
            {
                IAPIProxySettingsEndpoint endPoint = swaggerDocDictionaryEntry.Key;
                Log.Debug("processing {0}", endPoint.GetUrl());
                string methodNameAppend = string.Empty;
                if (endPoint.GetAppendAsyncToMethodName())
                {
                    methodNameAppend = "Async";
                }

                string result = swaggerDocDictionaryEntry.Value;
                SwaggerParser parser = new SwaggerParser();
                IProxyDefinition proxyDefinition = parser.ParseDoc(result, (RESTApiProxySettingsEndPoint)endPoint);
                string endPointString = string.Format("http://{0}{1}", proxyDefinition.Host, proxyDefinition.BasePath);
                if (!endPointString.EndsWith("/"))
                {
                    endPointString = string.Format(endPointString + "{0}", "/");
                }

                swaggerServiceDefinition.EndPoint = endPointString;
                List<string> proxies = proxyDefinition.Operations.Select(i => i.ProxyName).Distinct().ToList();
                /* Write interface for each proxy in proxies */
                foreach (string proxy in proxies)
                {
                    StringBuilder interfaceStringBuilder = CreateInterfaceStringBuilderForProxy(proxyDefinition, proxy, endPoint, methodNameAppend);
                    Log.Debug("created interfaceStringBuilder for {0}", proxy);
                    sourceStringList.Add(interfaceStringBuilder.ToString());
                }

                /* Write class for each proxy in proxies */
                foreach (string proxy in proxies)
                {
                    string className = SwaggerParser.FixTypeName(proxy) + "WebProxy";
                    swaggerServiceDefinition.ProxyClasses.Add(className);
                    StringBuilder classStringBuilder = CreateProxyStringBuilderForProxy(proxyDefinition, proxy, endPoint, methodNameAppend);
                    Log.Debug("created classStringBuilder for {0}", proxy);
                    sourceStringList.Add(classStringBuilder.ToString());
                }

                // Model Classes
                foreach (ClassDefinition classDefinition in proxyDefinition.ClassDefinitions)
                {
                    StringBuilder modelClassStringBuilder = WriteClassDefinitionToStringBuilder(classDefinition, endPoint);
                    sourceStringList.Add(modelClassStringBuilder.ToString());
                }
            }

            swaggerServiceDefinition.SourceStrings = sourceStringList.ToArray<string>();
            return swaggerServiceDefinition;
        }

        private static RESTServiceDefinition ProcessSwaggerDocuments()
        {
            Log.Debug("starting ProcessSwaggerDocuments()");
            return ProcessSwaggerDocuments("Admin", "password", "tenant1");
        }

        public static string GetDefaultValue(Parameter x)
        {
            if (!x.Type.IsNullableType && x.CollectionFormat != "multi" && x.Type.EnumValues != null && x.Type.EnumValues.Any())
            {
                return string.Format("{0}.{1}", GetDefaultType(x), x.Type.EnumValues.FirstOrDefault());
            }

            return "null";
        }

        public static RESTApiProxySettings GetSettings(string path)
        {
            using (FileStream settingStream = File.OpenRead(path))
            {
                StreamReader streamReader = new StreamReader(settingStream);
                var value = streamReader.ReadToEnd();
                return JsonConvert.DeserializeObject<RESTApiProxySettings>(value);
            }
        }
    }
}
